### ISSUE TITLE
Ubuntu detection broken with 'distro>=1.7.0'

### ISSUE THREAD

**Issue by cottsay** [ros-infrastructure/rospkg#241]
```
$ python3 os_detect.py
OS Name:     ubuntu
OS Version:  20.04
OS Codename: focal
$ pip3 install -q --user -U 'distro>=1.7.0'
$ python3 os_detect.py
OS Name:     ubuntu
OS Version:  20.04
OS Codename: Focal Fossa
```

First reported as ros-infrastructure/rosdep#779

**Comment by ruffsl** [comment]
Yea, I'm also seeing some inconsistencies with rosdep due to this issue in rospkg:
```
docker run -it --rm ros:galactic-ros-core-focal
root@81801f21a9f1:/# python3
Python 3.8.10 (default, Nov 26 2021, 20:14:08)
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from rospkg.os_detect import OsDetect
>>> print(OsDetect().get_codename())
focal
```

```
docker run -it --rm ros:rolling-ros-core-jammy
root@4b0460b1d500:/# python3
Python 3.9.10 (main, Jan 16 2022, 17:12:18)
[GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from rospkg.os_detect import OsDetect
>>> print(OsDetect().get_codename())
Jammy Jellyfish
```

**Comment by ruffsl** [comment]
Looks like the current call used here is deprecated:

https://github.com/ros-infrastructure/rospkg/blob/822140a7fa0147b11e1e130e79145b9d27064701/src/rospkg/os_detect.py#L142 


```
docker run -it --rm ros:rolling-ros-core-jammy
root@90bc489b4183:/# python3
Python 3.9.10 (main, Jan 16 2022, 17:12:18)
[GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import distro
>>> distro.linux_distribution(full_distribution_name=0)
<stdin>:1: DeprecationWarning: distro.linux_distribution() is deprecated. It should only be used as a compatibility shim with Python's platform.linux_distribution(). Please use distro.id(), distro.version() and distro.name() instead.
('ubuntu', '22.04', 'Jammy Jellyfish')
```

Seems like we should be using the codename function:

https://github.com/python-distro/distro/blob/5d228805b91b4e2363703cb30185156645be433d/src/distro/distro.py#L174-L175

https://github.com/python-distro/distro/blob/5d228805b91b4e2363703cb30185156645be433d/src/distro/distro.py#L441

```
>>> distro.codename()
'jammy'
```

**Comment by ros-discourse** [comment]
This issue has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/new-packages-for-ros-2-rolling-ridley/24431/2

**Comment by ruffsl** [comment]
Looks like the issue mentioned here bit us

https://github.com/python-distro/distro/issues/238

And the fix from this commit hasn't landed in jammy yet

https://github.com/python-distro/distro/commit/a08f31b676bd74995157aee566bb34d871ea89c0

We should still probably migrate from the deprecated shim.

**Comment by nuclearsandwich** [comment]
> Looks like the issue mentioned here bit us
> 
> [python-distro/distro#238](https://github.com/python-distro/distro/issues/238)

The regression was reportedly introduced in 1.4.0 which is the version in Ubuntu Focal[1] and the fix is in the recently released v1.7.0 tag and _is_ released in Ubuntu Jammy[2]. Is it possible that our "preferred' behavior is the incompatible (which would not make intuitive sense since we were using distro as a drop-in replacement) but if changes in Ubuntu conspired to make things seem continuous (or if we subtly changed our own expectations without noticing) then I could see this happening. I was looking at the immediate patch and haven't broadened out to review the codename's use downstream.

https://github.com/ros-infrastructure/rospkg/pull/248 is open as a draft.

[1]: https://packages.ubuntu.com/focal/python3-distro
[2]: https://packages.ubuntu.com/jammy/python3-distro

**Comment by nuclearsandwich** [comment]
Okay so as I read through rospkg sources we've always expected the third tuple member to be what I'll call the symbol or token codename rather than the full release nickname. That is, `focal` not `Focal Fossa`, `bionic` not `Bionic Beaver` all the way back to `lucid` not `Lucid Lynx`. The evidence for this is in the test code which establishes mocks with that pattern:

https://github.com/ros-infrastructure/rospkg/blob/2a437e834d63c42ee8782f85dd7ce8eb1578507b/test/test_rospkg_os_detect.py#L113-L116

and in local container testing I get the following results

#### Ubuntu 10.04 Lucid
*distro is not installed because pip is not packaged and also python 3.1....*
```
❯ lxc exec lucid bash
# python3
Python 3.1.2 (r312:79147, Oct 23 2012, 20:07:42)
[GCC 4.4.3] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import platform; platform.linux_distribution()
('Ubuntu', '10.04', 'lucid')
```

#### Ubuntu 16.04 Xenial with distro 1.6.0
*distro 1.7 does not install on Xenial due to typing syntax incompatibilities.*

```
❯ lxc exec xenial bash
# python3
Python 3.5.2 (default, Jan 26 2021, 13:30:48)
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import platform; import distro; (platform.linux_distribution(), distro.linux_distribution())
(('Ubuntu', '16.04', 'xenial'), ('Ubuntu', '16.04', 'xenial'))
```



#### Ubuntu 18.04 Bionic with distro 1.6.0 and 1.7.0

```
❯ lxc exec bionic bash
# pip3 install distro==1.6.0
# python3
Python 3.6.9 (default, Dec  8 2021, 21:08:43)
[GCC 8.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import platform; import distro; (platform.linux_distribution(), distro.linux_distribution())
(('Ubuntu', '18.04', 'bionic'), ('Ubuntu', '18.04', 'bionic'))
# pip3 install distro==1.7.0
# python3
Python 3.6.9 (default, Dec  8 2021, 21:08:43)
[GCC 8.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import platform; import distro; (platform.linux_distribution(), distro.linux_distribution())
(('Ubuntu', '18.04', 'bionic'), ('Ubuntu', '18.04', 'Bionic Beaver'))
```

Which is all to say either I am reading the bug report in https://github.com/python-distro/distro/issues/238 or the report has something backwards because the evidence in front of me is that the behavior between 1.4.0 and 1.6.0 matched that of the platform.linux_distribution function.

However, across all tested versions of the distro module the `distro.codename()` function returns what we want and so the patch in #248 should resolve our issues on all currently supported platforms.
